### PR TITLE
Watch plugin for toggling --fix

### LINKED ISSUE #43: Watch Mode Plugin: jest-runner-eslint/watch
Combining https://github.com/facebook/jest/issues/5212 and the new Jest Watch Plugins API, it would be really cool to have:

```json
{
  "jest": {
    "runner": "jest-runner-eslint",
    "watchPlugins": [
      "jest-runner-eslint/watch"
    ]
  }
}
```

Which, when an ESLint error is reported, exposes an option in watch mode:

```
Press F to run eslint --fix
```

I think it makes sense for it to live in this repo, but it could equally be a standalone repo.

I'd be happy to work on a PR if it is desired here 😄 

Related: #10

### PR DESCRIPTION
**Pull request by rogeliog**
Fixes #43 

I was originally thinking of a more fully featured watch plugin https://github.com/jest-community/jest-runner-eslint/issues/43#issuecomment-395504661, but I was not happy with that UX.

Given that `--fix` is the most common request, I think it makes sense to start with a `--fix` plugin, and if need be we can work on another one later.

fix

- This PR depends on #49, so disregard the babel@7 stuff, I'll rebase once that is merged.
- I intentionally left the plugin in `jest-runner-eslint/watch-fix` instead of `jest-runner-eslint/watch` to leave room for more plugins.
- The way the plugin and the runner communicate is through `configOverrides`, I would love to get some feedback on this. I'm not a fan of it, but we need a way that they can both access shared state... Other ideas are more than welcome.
- One limitation or known issue is: We don't know initially what the current value of `--fix` is in the config. Given this, I decided to show a different message until the user enables the toggling. I think that the UX is still fine.

### CODE REVIEW

File: src/watchFixPlugin/index.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  getUsageInfo() {
+    return {
+      key: this._key,
+      prompt: this._getPrompt(),
+    };
+  }
+
+  async _showLoading(fix) {
```
Comment (SimenB) [inline src/watchFixPlugin/index.js]: why do we need this?

File: src/runner/__tests__/runESLint.test.js [hunk 2]
```diff
@@ -24,7 +24,7 @@ const runESLintRunnerWithMockedEngine = options => {
   });
   const runESLint = require('../runESLint');
 
-  return runESLint(options.runESLint);
+  return runESLint(Object.assign({ extraOptions: {} }, options.runESLint));
```
Comment (SimenB) [inline src/runner/__tests__/runESLint.test.js]: object spread?

File: README.md [hunk 3]
```diff
@@ -172,6 +193,8 @@ module.exports = {
 }
 ```
 
+###
```
Comment (SimenB) [inline README.md]: ?

File: src/runner/runESLint.js [hunk 4]
```diff
@@ -18,12 +18,18 @@ const runESLint = ({ testPath, config }) => {
   }
 
   const { CLIEngine } = getLocalESLint(config);
-  const { cliOptions } = getESLintOptions(config);
-  const cli = new CLIEngine(
-    Object.assign({}, cliOptions, {
-      fix: getComputedFixValue(cliOptions),
-    }),
+  const { cliOptions: baseCliOptions } = getESLintOptions(config);
+  const cliOptions = Object.assign(
```
Comment (SimenB) [inline src/runner/runESLint.js]: object spread

File: src/watchFixPlugin/__tests__/index.test.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+jest.doMock('chalk', () => new chalk.constructor({ enabled: false }));
+
+jest.useFakeTimers();
+
+let WatchFixPlugin;
+let configOverrides;
+
+const runAndAwait = async plugin => {
+  const runPromise = plugin.run();
+  await jest.runTimersToTime(700);
```
Comment (SimenB) [inline src/watchFixPlugin/__tests__/index.test.js]: we have `runAllPendingTimers` and `runAllTimers` instead of hardcoding the values

Comment (SimenB) [reply src/watchFixPlugin/__tests__/index.test.js]: They also don't return promises, so you can drop the `await`

File: src/watchFixPlugin/index.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+const wait = time => new Promise(res => setTimeout(res, time));
+
+class ESLintWatchFixPlugin {
+  constructor({ stdout, config }) {
+    this._stdout = stdout;
+    this._key = config.key || 'F';
+  }
+
+  // eslint-disable-next-line class-methods-use-this
```
Comment (SimenB) [inline src/watchFixPlugin/index.js]: needed? It _does_ use `this`

File: src/watchFixPlugin/index.js (hunk 1 above)
Comment (SimenB) [reply src/watchFixPlugin/index.js]: It looks slick, but does it do anything? Other than delay for a tiny bit

File: README.md (hunk 3 above)
Comment (rogeliog) [reply README.md]: 🤷‍♂️

File: src/watchFixPlugin/index.js (hunk 1 above)
Comment (rogeliog) [reply src/watchFixPlugin/index.js]: The only reason was to provide an indicator and feedback of what is happening, but I guess you are right... none of the other Jest keys does that, I just removed it.

**Comment by rogeliog** [comment]
@ljharb Any thoughts?

File: package.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "jest-runner-eslint",
   "version": "0.6.0",
-  "main": "build/index.js",
+  "main": "build/runner/index.js",
```
Comment (ljharb) [inline package.json]: this can just be build/runner

File: src/watchFixPlugin/index.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+  getUsageInfo() {
+    return {
+      key: this._key,
+      prompt: this._getPrompt(),
+    };
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  _getPrompt() {
```
Comment (ljharb) [inline src/watchFixPlugin/index.js]: why can’t this be a closed over function instead of a fake private method?

Comment (rogeliog) [reply src/watchFixPlugin/index.js]: 👍

File: src/watchFixPlugin/index.js [hunk 9]
```diff
@@ -15,23 +15,22 @@ class ESLintWatchFixPlugin {
   }
 
   getUsageInfo() {
+    const getPrompt = () => {
```
Comment (ljharb) [inline src/watchFixPlugin/index.js]: if this is defined at module level, then it should be a bit faster.

**Comment by SimenB** [comment]
(we should do #54 before releasing this, BTW)

**Comment by rogeliog** [comment]
@SimenB Perfect!

**Pull request merged**